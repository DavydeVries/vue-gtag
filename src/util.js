export function loadScript(url) {
  return new Promise((resolve, reject) => {
    var head = document.head || document.getElementsByTagName("head")[0];
    const script = document.createElement("script");
    script.async = true;
    script.src = url;
    script.setAttribute("rel", "preconnect");
    script.charset = "utf-8";

    head.appendChild(script);

    script.onload = resolve;
    script.onerror = reject;
  });
}

export function warn(msg, err) {
  if (typeof console !== "undefined") {
    console.warn("[vue-gtag] " + msg);

    if (err && err.stack) {
      console.warn(err.stack);
    }
  }
}

export function isFn(item) {
  return typeof item === "function";
}

export function isObject(item) {
  return item && typeof item === "object" && !Array.isArray(item);
}

export function execFunc(func, ...args) {
  return isFn(func) && func(...args);
}

export function mergeDeep(target, ...sources) {
  if (!sources.length) return target;
  const source = sources.shift();

  if (isObject(target) && isObject(source)) {
    for (const key in source) {
      if (isObject(source[key])) {
        if (!target[key]) Object.assign(target, { [key]: {} });
        mergeDeep(target[key], source[key]);
      } else {
        Object.assign(target, { [key]: source[key] });
      }
    }
  }

  return mergeDeep(target, ...sources);
}
