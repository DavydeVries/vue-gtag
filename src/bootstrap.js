import { warn, loadScript } from "./util";
import { getOptions } from "../src/install";
import pageTracker from "./page-tracker";

export default function() {
  if (typeof document === "undefined" || typeof window === "undefined") {
    return;
  }

  const {
    enabled,
    globalObjectName,
    config,
    pageTrackerEnabled
  } = getOptions();

  if (!enabled) {
    window[`ga-disable-${config.id}`] = true;
  }

  window.dataLayer = window.dataLayer || [];
  window[globalObjectName] = function() {
    window.dataLayer.push(arguments);
  };

  window[globalObjectName]("js", new Date());
  window[globalObjectName]("config", config.id, config.params);

  if (pageTrackerEnabled) {
    pageTracker();
  }

  return loadScript(`https://www.googletagmanager.com/gtag/js?id=${config.id}`)
    .then(() => window[globalObjectName])
    .catch(error => {
      warn("Ops! Something happened and gtag.js couldn't be loaded", error);
      return error;
    });
}
