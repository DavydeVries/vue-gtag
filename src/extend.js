import { getVue } from "./install";
import api from "./api";

export default () => {
  const Vue = getVue();

  Vue.$gtag = Vue.prototype.$gtag = api;
};
