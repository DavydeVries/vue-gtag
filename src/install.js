import { mergeDeep } from "./util";
import extend from "./extend";
import bootstrap from "./bootstrap";

let Vue;
let Router;
let options = {
  enabled: true,
  bootstrap: true,
  globalObjectName: "gtag",
  pageTrackerTemplate: null,
  pageTrackerEnabled: true,
  pageTrackerScreenviewEnabled: false,
  config: {
    id: null,
    params: {
      send_page_view: true
    }
  }
};

export const getOptions = () => options;
export const getVue = () => Vue;
export const getRouter = () => Router;

export function install(_Vue, _options = {}, _Router) {
  Vue = _Vue;
  Router = _Router;
  options = mergeDeep(options, _options);

  extend();

  if (options.bootstrap) {
    bootstrap();
  }
}
