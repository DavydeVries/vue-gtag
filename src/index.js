import { install } from "./install";
import _bootstrap from "./bootstrap";

export default install;

export const bootstrap = _bootstrap;
