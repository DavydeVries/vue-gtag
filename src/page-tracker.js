import { getRouter, getOptions } from "./install";
import { isFn, warn } from "./util";
import config from "./api/config";
import screenview from "./api/screenview";

export const trackPage = (to, from) => {
  if (to.path === from.path) {
    return;
  }

  let template;
  const {
    pageTrackerTemplate,
    pageTrackerScreenviewEnabled,
    appName
  } = getOptions();

  const customTemplate = isFn(pageTrackerTemplate)
    ? pageTrackerTemplate(to, from)
    : null;

  if (customTemplate) {
    template = customTemplate;
  } else if (pageTrackerScreenviewEnabled) {
    template = {
      app_name: appName,
      screen_name: to.name
    };
  } else {
    template = {
      page_title: to.name,
      page_path: to.path,
      page_location: window.location.href
    };
  }

  if (pageTrackerScreenviewEnabled && !template.app_name) {
    warn("To use the screenview, add the appName to the plugin options");
    return;
  }

  if (pageTrackerScreenviewEnabled && !template.screen_name) {
    warn("To use the screenview, name your routes");
    return;
  }

  if (pageTrackerScreenviewEnabled) {
    screenview(template);
    return;
  }

  config(template);
};

export default () => {
  const Router = getRouter();

  if (!Router) {
    return;
  }

  Router.onReady(() => {
    Router.afterEach(trackPage);
  });
};
