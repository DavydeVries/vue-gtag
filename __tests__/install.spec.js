import { install, getOptions } from "../src/install";
import bootstrap from "../src/bootstrap";
import extend from "../src/extend";

jest.mock("../src/extend");
jest.mock("../src/bootstrap");

describe("install", () => {
  it("should have default options", () => {
    expect(getOptions()).toMatchSnapshot();
  });

  it("should call extend", () => {
    install();
    expect(extend).toHaveBeenCalled();
  });

  it("should call bootstrap", () => {
    install();
    expect(bootstrap).toHaveBeenCalled();
  });
});
