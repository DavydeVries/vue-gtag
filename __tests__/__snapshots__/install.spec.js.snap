// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`install should have default options 1`] = `
Object {
  "bootstrap": true,
  "config": Object {
    "id": null,
    "params": Object {
      "send_page_view": true,
    },
  },
  "enabled": true,
  "globalObjectName": "gtag",
  "pageTrackerEnabled": true,
  "pageTrackerScreenviewEnabled": false,
  "pageTrackerTemplate": null,
}
`;
